src/web/components/ClientSideTokenGeneration/CstgAddDialog.tsx

```diff
@@ -0,0 +1,136 @@
+
+import {
+  deduplicateStrings,
+  formatStringsWithSeparator,
+  separateStringsList,
+} from '../../utils/textHelpers';
+import { Dialog } from '../Core/Dialog';
+import { RootFormErrors } from '../Input/FormError';
+import { MultilineTextInput } from '../Input/MultilineTextInput';
+import { StyledCheckbox } from '../Input/StyledCheckbox';
+import {
+  AddCstgValuesFormProps,
+  CstgValueType,
+  extractTopLevelDomain,
+  validateAppId,
+} from './CstgHelper';
+
+import './CstgAddDialog.scss';
+
+type CstgAddDialogProps = Readonly<{
+  onAddCstgValues: (newValuesFormatted: string[], deleteExistingList: boolean) => Promise<string[]>;
+  onOpenChange: () => void;
+  existingCstgValues: string[];
+  cstgValueType: CstgValueType;
+  getUniqueValues: (
+    newCstgValues: string[],
+    existingCstgValues: string[],
+    deleteExistingList: boolean
+  ) => string[];
+  addInstructions: string;
+}>;
+
+function CstgAddDialog({
+  onAddCstgValues,
+  onOpenChange,
+  existingCstgValues,
+  cstgValueType,
+  getUniqueValues,
+  addInstructions,
+}: CstgAddDialogProps) {
+  const [deleteExistingList, setDeleteExistingList] = useState<boolean>(false);
+  const formMethods = useForm<AddCstgValuesFormProps>();
+  const {
+    handleSubmit,
+    setError,
+    formState: { errors },
+  } = formMethods;
+
+  const handleError = (message: string) => {
+    setError('root.serverError', {
+      type: '400',
+      message,
+    });
+  };
+
+  const onSubmit = async (formData: AddCstgValuesFormProps) => {
+    const newCstgValues = getUniqueValues(
+      separateStringsList(formData.cstgValues),
+      existingCstgValues,
+      deleteExistingList
+    );
+    if (newCstgValues.length === 0) {
+      handleError(`The ${cstgValueType}s entered already exist.`);
+    } else if (cstgValueType === CstgValueType.Domain) {
+      newCstgValues.forEach((newDomain, index) => {
+        newCstgValues[index] = extractTopLevelDomain(newDomain);
+      });
+      // filter for uniqueness (e.g. 2 different domains entered could have the same root-level domain)
+      const dedupedRootLevelDomains = deduplicateStrings(newCstgValues);
+      const invalidDomains = await onAddCstgValues(dedupedRootLevelDomains, deleteExistingList);
+      if (invalidDomains.length > 0) {
+        handleError(
+          `The domains entered are invalid root-level domains: ${formatStringsWithSeparator(
+            invalidDomains
+          )}`
+        );
+      }
+    } else if (cstgValueType === CstgValueType.MobileAppId) {
+      const invalidAppIds: string[] = [];
+      newCstgValues.forEach((newAppId) => {
+        if (!validateAppId(newAppId)) {
+          invalidAppIds.push(newAppId);
+        }
+      });
+      if (invalidAppIds.length > 0) {
+        handleError(
+          `The mobile app ids entered are invalid: ' ${formatStringsWithSeparator(invalidAppIds)}`
+        );
+      } else {
+        await onAddCstgValues(newCstgValues, deleteExistingList);
+      }
+    }
+  };
```

Comment:
This function may look a little funny between domain and app id - the reason is because currently, domain validation is on the backend and app id validation is on the front end.  Domains also need to be deduped a second time to filter out the case that two entered domains lead to the same +1 domain (ex. photos.google.com and drive.google.com both -> google.com)

Reply:
Another way to approach this would be to inject the validation behaviour - either by providing a validator function or else just leaving validation to the `onAddCstgValues` handler. But ... we're expecting some re-work on this anyway when we do the UX enhancement for the app IDs, so I think this is fine for now.